lib/ansible/module_utils/azure_rm_common.py

```diff
@@ -767,18 +780,65 @@ def create_default_securitygroup(self, resource_group, location, security_group_
         session.verify = False
 
+    def get_api_profile(self, client_type_name, api_profile_name):
+        profile_all_clients = AZURE_API_PROFILES.get(api_profile_name)
+
+        if not profile_all_clients:
+            raise KeyError("unknown Azure API profile: {0}".format(api_profile_name))
+
+        profile_raw = profile_all_clients.get(client_type_name, None)
+
+        if not profile_raw:
+            self.module.warn("Azure API profile {0} does not define an entry for {1}".format(api_profile_name, client_type_name))
+
+        if isinstance(profile_raw, dict):
+            if not profile_raw.get('default_api_version'):
+                raise KeyError("Azure API profile {0} does not define 'default_api_version'".format(api_profile_name))
+            return profile_raw
+
+        # wrap basic strings in a dict that just defines the default
+        return dict(default_api_version=profile_raw)
+
     def get_mgmt_svc_client(self, client_type, base_url=None, api_version=None):
         self.log('Getting management service client {0}'.format(client_type.__name__))
         self.check_client_version(client_type)
-        if api_version:
-            client = client_type(self.azure_credentials,
-                                 self.subscription_id,
-                                 api_version=api_version,
-                                 base_url=base_url)
-        else:
-            client = client_type(self.azure_credentials,
-                                 self.subscription_id,
-                                 base_url=base_url)
+
+        client_argspec = inspect.getargspec(client_type.__init__)
+
+        client_kwargs = dict(credentials=self.azure_credentials, subscription_id=self.subscription_id, base_url=base_url)
+
+        api_profile_dict = {}
+
+        if self.api_profile:
+            api_profile_dict = self.get_api_profile(client_type.__name__, self.api_profile)
+
+        if not base_url:
+            # most things are resource_manager, don't make everyone specify
+            base_url = self._cloud_environment.endpoints.resource_manager
+
+        # unversioned clients won't accept profile; only send it if necessary
+        # clients without a version specified in the profile will use the default
+        if api_profile_dict and 'profile' in client_argspec.args:
+            client_kwargs['profile'] = api_profile_dict
+
+        # If the client doesn't accept api_version, it's unversioned.
+        # If it does, favor explicitly-specified api_version, fall back to api_profile
+        if 'api_version' in client_argspec.args:
+            profile_default_version = api_profile_dict.get('default_api_version', None)
+            if api_version or profile_default_version:
+                client_kwargs['api_version'] = api_version or profile_default_version
+
+        client = client_type(**client_kwargs)
+
+        # FUTURE: remove this once everything exposes models directly (eg, containerinstance)
```

Comment:
ContainerInstance doesn't support profile, it's why it doesn't have a `models` method

Reply:
so again, as with api_version, is there a reason that it shouldn't have models method just for consistency?

Reply:
`models` in multi-api packages takes an api-version in parameter and return the models for this ApiVersion.
On a single api package, this doesn't make sense to have an api-version parameter. If I do `models`, the most logical would be to have no parameter. Technically you could do:
```python
try:
   models = client.models(api_version)
except TypeError: # Does not support api_version parameter
   models = client.models()
```
@zikalino @nitzmahone Would this help?